src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -316,9 +397,44 @@ private string FormatEnumValueToString(string value, JavaScriptEncoder? encoder)
             return converted;
         }
 
-        internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        private static string[] SplitFlagsEnum(string value)
+        {
+            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
+            return value.Split(
+#if NETCOREAPP
+                ValueSeparator
+#else
+                new string[] { ValueSeparator }, StringSplitOptions.None
+#endif
+                );
+        }
+
+#if NETCOREAPP
+        private T ReadEnumFromString(ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            if (TryParseEnumCore(ref reader, options, out T value))
+            {
+                return value;
+            }
+
+            return ReadEnumUsingNamingPolicy(reader.GetString());
+        }
+#else
+        private T ReadEnumFromString(ref Utf8JsonReader reader, JsonSerializerOptions options)
         {
+            string? enumString = reader.GetString();
+            if (TryParseEnumCore(enumString, options, out T value))
+            {
+                return value;
+            }
+
+            return ReadEnumUsingNamingPolicy(enumString);
+        }
+#endif
```

Comment:
Seems like this could work?

```suggestion
        private T ReadEnumFromString(ref Utf8JsonReader reader, JsonSerializerOptions options)
        {
#if NETCOREAPP
            if (TryParseEnumCore(ref reader, options, out T value))
#else
            string? enumString = reader.GetString();
            if (TryParseEnumCore(enumString, options, out T value))
#endif
            {
                return value;
            }

#if NETCOREAPP
            return ReadEnumUsingNamingPolicy(reader.GetString());
#else 
            return ReadEnumUsingNamingPolicy(enumString);
#endif
        }
```